docs/design/secrets.md

```diff
@@ -0,0 +1,486 @@
+component that could regenerate expired secrets.
+
+## Deferral: Consuming secrets as environment variables
+
+Some images will expect to receive configuration items as environment variables instead of files.
+We should consider what the best way to allow this is; there are a few different options:
+
+1.  Force the user to adapt files into environment variables.  Users can store secrets that need to
+    be presented as environment variables in a format that is easy to consume from a shell:
+
+        $ cat /etc/secrets/my-secret.txt
+        export MY_SECRET_ENV=MY_SECRET_VALUE
+
+    The user could `source` the file at `/etc/secrets/my-secret` prior to executing the command for
+    the image either inline in the command or in an init script, 
+
+2.  Give secrets an attribute that allows users to express the intent that the platform should
+    generate the above syntax in the file used to present a secret.  The user could consume these
+    files in the same manner as the above option.
+
+3.  Give secrets attributes that allow the user to express that the secret should be presented to
+    the container as an environment variable.  The container's environment would contain the
+    desired values and the software in the container could use them without accomodation the
+    command or setup script.
+
+For our initial work, we will treat all secrets as files to narrow the problem space.  There will
+be a future proposal that handles exposing secrets as environment variables.
+
+## Flow analysis of secret data with respect to the API server
+
+There are two fundamentally different use-cases for access to secrets:
+
+1.  CRUD operations on secrets by their owners
+2.  Read-only access to the secrets needed for a particular node by the kubelet
+
+### Use-Case: CRUD operations by owners
+
+In use cases for CRUD operations, the user experience for secrets should be no different than for
+other API resources.
+
+#### Data store backing the REST API
+
+The data store backing the REST API should be pluggable because different cluster operators will
+have different preferences for the central store of secret data.  Some possibilities for storage:
+
+1.  An etcd collection alongside the storage for other API resources
+2.  A collocated [HSM](http://en.wikipedia.org/wiki/Hardware_security_module)
+3.  An external datastore such as an external etcd, RDBMS, etc.
+
+#### Size limit for secrets
+
+There should be a size limit for secrets in order to:
+
+1.  Prevent DOS attacks against the API server
+2.  Allow kubelet implementations that prevent secret data from touching the node's filesystem
+
+The size limit should satisfy the following conditions:
+
+1.  Large enough to store common artifact types (encryption keypairs, certificates, small
+    configuration files)
+2.  Small enough to avoid large impact on node resource consumption (storage, RAM for tmpfs, etc)
+
+To begin discussion, we propose an initial value for this size limit of **1MB**.
```

Comment:
The largest i can think of is a PEM chain or a kerb key tab.  I do however thing folks will sign and encrypt json blobs for this.  The naive secret implementation uses etcd so that imposes one upper bound.

----- Original Message -----
> > +3.  An external datastore such as an external etcd, RDBMS, etc.
> > +
> > +#### Size limit for secrets
> > +
> > +There should be a size limit for secrets in order to:
> > +
> > +1.  Prevent DOS attacks against the API server
> > +2.  Allow kubelet implementations that prevent secret data from touching
> > the node's filesystem
> > +
> > +The size limit should satisfy the following conditions:
> > +
> > +1.  Large enough to store common artifact types (encryption keypairs,
> > certificates, small
> > +    configuration files)
> > +2.  Small enough to avoid large impact on node resource consumption
> > (storage, RAM for tmpfs, etc)
> > +
> > +To begin discussion, we propose an initial value for this size limit of
> > **1MB**.
> 
> Do we have examples of "large" secrets to get an idea of whether this is
> reasonable or not?  All secrets I can think of are a few kb max (ssh keys
> typically).
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/GoogleCloudPlatform/kubernetes/pull/4126/files#r24807159